File: src/main/java/com/vermouthx/stocker/utils/StockerTableModelUtil.java
package com.vermouthx.stocker.utils;

import com.vermouthx.stocker.enums.StockerTableColumn;

import javax.swing.table.DefaultTableModel;
import java.util.Objects;

public final class StockerTableModelUtil {

    public static int existAt(DefaultTableModel tableModel, String code) {
        int codeCol = tableModel.findColumn(StockerTableColumn.SYMBOL.name());
        if (codeCol < 0) codeCol = 0;
        for (int i = 0; i < tableModel.getRowCount(); i++) {
            Object cell = tableModel.getValueAt(i, codeCol);
            if (cell != null && code != null && code.equals(cell.toString())) {
                return i;
            }
        }
        return -1;
    }

    /**
     * Update a cell only when the new value differs from the existing one, firing
     * {@code fireTableCellUpdated} exactly when a change happened. No-op for unknown columns.
     */
    public static void setIfChanged(DefaultTableModel model, int row, int col, Object newValue) {
        if (col < 0 || col >= model.getColumnCount() || row < 0 || row >= model.getRowCount()) {
            return;
        }
        Object existing = model.getValueAt(row, col);
        if (!Objects.equals(existing, newValue)) {
            model.setValueAt(newValue, row, col);
            model.fireTableCellUpdated(row, col);
        }
    }

    /** Resolve a model column index by {@link StockerTableColumn} enum, returning -1 if absent. */
    public static int colOf(DefaultTableModel model, StockerTableColumn column) {
        return model.findColumn(column.name());
    }

    private StockerTableModelUtil() {
    }
}

File: src/main/kotlin/com/vermouthx/stocker/utils/StockerTableModelUtil.kt
package com.vermouthx.stocker.utils

import com.vermouthx.stocker.enums.StockerTableColumn
import javax.swing.table.DefaultTableModel

object StockerTableModelUtil {

    @JvmStatic
    fun existAt(tableModel: DefaultTableModel, code: String?): Int {
        if (code == null) return -1
        val codeCol = tableModel.findColumn(StockerTableColumn.SYMBOL.name).let { if (it < 0) 0 else it }
        for (i in 0 until tableModel.rowCount) {
            val cell = tableModel.getValueAt(i, codeCol)
            if (cell != null && code == cell.toString()) return i
        }
        return -1
    }

    /**
     * Update a cell only when the new value differs from the existing one. The single
     * cell-updated event comes from DefaultTableModel.setValueAt itself — don't fire a
     * second one here. No-op for unknown columns.
     */
    @JvmStatic
    fun setIfChanged(model: DefaultTableModel, row: Int, col: Int, newValue: Any?) {
        if (col < 0 || col >= model.columnCount || row < 0 || row >= model.rowCount) return
        val existing = model.getValueAt(row, col)
        if (existing != newValue) {
            model.setValueAt(newValue, row, col)
        }
    }

    /**
     * One-pass code→row index for batch updates: looking each quote up via [existAt]
     * made a refresh tick O(rows²). Same symbol-column fallback as [existAt].
     */
    @JvmStatic
    fun rowIndexByCode(tableModel: DefaultTableModel): MutableMap<String, Int> {
        val codeCol = tableModel.findColumn(StockerTableColumn.SYMBOL.name).let { if (it < 0) 0 else it }
        val index = HashMap<String, Int>(tableModel.rowCount * 2)
        for (i in 0 until tableModel.rowCount) {
            val cell = tableModel.getValueAt(i, codeCol) ?: continue
            index[cell.toString()] = i
        }
        return index
    }

    /** Resolve a model column index by [StockerTableColumn] enum, returning -1 if absent. */
    @JvmStatic
    fun colOf(model: DefaultTableModel, column: StockerTableColumn): Int =
        model.findColumn(column.name)
}
